pkg/metrics/metrics.jsx

```diff
@@ -917,163 +918,141 @@ class MetricsHour extends React.Component {
+        const check = () => {
+            if (cond()) {
+                objects.forEach(o => o.removeEventListener("changed", check));
+                resolve();
+            }
+        };
+        objects.forEach(o => o.addEventListener("changed", check));
+        check();
+    });
+};
+
+const PCPConfigDialog = ({
+    firewalldRequest,
+    needsLogout, setNeedsLogout,
+    s_pmlogger, s_pmproxy, s_redis, s_redis_server
+}) => {
+    const Dialogs = useDialogs();
+    const dialogInitialProxyValue = runningService(s_pmproxy) && (runningService(s_redis) || runningService(s_redis_server));
     const [dialogError, setDialogError] = useState(null);
-    const [dialogLoggerValue, setDialogLoggerValue] = useState(false);
-    const [dialogProxyValue, setDialogProxyValue] = useState(null);
-    const [dialogInitialProxyValue, setDialogInitialProxyValue] = useState(null);
+    const [dialogLoggerValue, setDialogLoggerValue] = useState(runningService(s_pmlogger));
+    const [dialogProxyValue, setDialogProxyValue] = useState(dialogInitialProxyValue);
     const [pending, setPending] = useState(false);
-    const [deferSaveOnInstall, setDeferSaveOnInstall] = useState(false);
 
-    const s_pmlogger = useObject(() => service.proxy("pmlogger.service"), null, []);
-    const s_pmproxy = useObject(() => service.proxy("pmproxy.service"), null, []);
-    // redis.service on Fedora/RHEL, redis-server.service on Debian/Ubuntu with an Alias=redis
-    const s_redis = useObject(() => service.proxy("redis.service"), null, []);
-    const s_redis_server = useObject(() => service.proxy("redis-server.service"), null, []);
-
-    useEvent(superuser, "changed");
-    useEvent(s_pmlogger, "changed");
-    useEvent(s_pmproxy, "changed");
-    useEvent(s_redis, "changed");
-    useEvent(s_redis_server, "changed");
-
-    let real_redis;
-    let redis_name;
-    if (s_redis_server.exists) {
-        real_redis = s_redis_server;
-        redis_name = "redis-server.service";
-    } else {
-        real_redis = s_redis;
-        redis_name = "redis.service";
-    }
-
-    debug("PCPConfig s_pmlogger.state", s_pmlogger.state, "needs logout", needsLogout, "deferSaveOnInstall", deferSaveOnInstall);
-    debug("PCPConfig s_pmproxy state", s_pmproxy.state, "redis exists", s_redis.exists, "state", s_redis.state, "redis-server exists", s_redis_server.exists, "state", s_redis_server.state);
-
-    if (!superuser.allowed)
-        return null;
+    const handleInstall = () => {
+    // when enabling services, install missing packages on demand
+        const missing = [];
+        if (dialogLoggerValue && !s_pmlogger.exists)
+            missing.push("cockpit-pcp");
+        if (dialogProxyValue && !(s_redis.exists || s_redis_server.exists))
+            missing.push("redis");
+        if (missing.length > 0) {
+            debug("PCPConfig: missing packages", JSON.stringify(missing), ", offering install");
+            Dialogs.close();
```

Comment:
This is a bit awkward -- it means that on installation we can't show enablement errors in the dialog any more. But the old code already did that, so not a regression, and not a blocker for this PR. Just spotting my own old sins :grin:

Reply:
Yeah, I noticed that too.

When packages need to be installed in order to carry out some action, we usually install them before opening the action dialog, like when installing Stratis before creating the first pool. There we go from the generic "Install" dialog to the "Create pool" dialog.

When package installation might or might not be needed, I think we need to announce that as part of the action dialog itself, and carry it out as part of the action. We do this for Subscription Manager, where Insights is installed when the checkbox is "on". The generic "Install" dialog is not used for this, see https://github.com/candlepin/subscription-manager-cockpit/blob/main/src/insights.jsx#L150 and https://github.com/candlepin/subscription-manager-cockpit/blob/main/src/insights.jsx#L78.

We should do this here as well, ideally.